### PR TITLE
jpa-hw done

### PR DESCRIPTION
**Pull request by BelitskyiAlexandr**

### CODE REVIEW

File: src/main/java/com/example/bookshop/service/impl/BookServiceImpl.java [hunk 1]
```diff
@@ -32,8 +32,22 @@ public List<BookDto> findAll() {
...
                 new EntityNotFoundException("Cannot find book by id:" + id));
         return bookMapper.toBookDto(book);
     }
+
+    @Override
+    public BookDto updateById(Long id, CreateBookRequestDto createBookRequestDto) {
+        bookRepository.findById(id).orElseThrow(() ->
+                new EntityNotFoundException("Cannot find book by id: " + id));
+        Book book = bookMapper.toModel(createBookRequestDto);
+        book.setId(id);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/service/impl/BookServiceImpl.java:45]:
```suggestion
       bookMapper.updateBookFromDto(createBookRequestDto, book);
```

File: src/main/resources/db/changelog/changes/01-create-book-table.yaml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+              - column:
+                  name: price
+                  type: bigint
+                  constraints:
+                    nullable: false
+              - column:
+                  name: description
+                  type: varchar(255)
+              - column:
+                  name: coverImage
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/01-create-book-table.yaml:41]:
```suggestion
                  name: cover_image
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Comment by BelitskyiAlexandr** [comment]
could you tell me if it is ok that i used CreateBookRequestDto to update the book instead of creating a new dto?

There is simply little logic in creating a new identical dto, where only the name is different. Is it necessary to somehow change name in this case?

**Comment by BelitskyiAlexandr** [comment]
There are some problems with my app. When I try to execute it, I get an exception:

```

  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/

 :: Spring Boot ::                (v3.3.2)

2024-08-27T17:40:38.081+03:00  INFO 6992 --- [           main] c.e.bookshop.SpringBootIntroApplication  : Starting SpringBootIntroApplication using Java 17.0.2 with PID 6992 (D:\dyst\proj\spring\spring-boot-intro\target\classes started by ADMIN in D:\dyst\proj\spring\spring-boot-intro)
2024-08-27T17:40:38.087+03:00  INFO 6992 --- [           main] c.e.bookshop.SpringBootIntroApplication  : No active profile set, falling back to 1 default profile: "default"
2024-08-27T17:40:39.400+03:00  INFO 6992 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Bootstrapping Spring Data JPA repositories in DEFAULT mode.
2024-08-27T17:40:39.496+03:00  INFO 6992 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Finished Spring Data repository scanning in 81 ms. Found 1 JPA repository interface.
2024-08-27T17:40:40.456+03:00  INFO 6992 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat initialized with port 8080 (http)
2024-08-27T17:40:40.479+03:00  INFO 6992 --- [           main] o.apache.catalina.core.StandardService   : Starting service [Tomcat]
2024-08-27T17:40:40.479+03:00  INFO 6992 --- [           main] o.apache.catalina.core.StandardEngine    : Starting Servlet engine: [Apache Tomcat/10.1.26]
2024-08-27T17:40:40.601+03:00  INFO 6992 --- [           main] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring embedded WebApplicationContext
2024-08-27T17:40:40.603+03:00  INFO 6992 --- [           main] w.s.c.ServletWebServerApplicationContext : Root WebApplicationContext: initialization completed in 2432 ms
2024-08-27T17:40:40.915+03:00  INFO 6992 --- [           main] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Starting...
2024-08-27T17:40:41.427+03:00  INFO 6992 --- [           main] com.zaxxer.hikari.pool.HikariPool        : HikariPool-1 - Added connection com.mysql.cj.jdbc.ConnectionImpl@45a1d057
2024-08-27T17:40:41.430+03:00  INFO 6992 --- [           main] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Start completed.
2024-08-27T17:40:42.540+03:00  INFO 6992 --- [           main] liquibase.changelog                      : Reading from bookshop_db.databasechangelog
2024-08-27T17:40:42.596+03:00  INFO 6992 --- [           main] liquibase.ui                             : Database is up to date, no changesets to execute
2024-08-27T17:40:42.600+03:00  INFO 6992 --- [           main] liquibase.changelog                      : Reading from bookshop_db.databasechangelog
2024-08-27T17:40:42.611+03:00  INFO 6992 --- [           main] liquibase.util                           : UPDATE SUMMARY
2024-08-27T17:40:42.612+03:00  INFO 6992 --- [           main] liquibase.util                           : Run:                          0
2024-08-27T17:40:42.612+03:00  INFO 6992 --- [           main] liquibase.util                           : Previously run:               1
2024-08-27T17:40:42.612+03:00  INFO 6992 --- [           main] liquibase.util                           : Filtered out:                 0
2024-08-27T17:40:42.613+03:00  INFO 6992 --- [           main] liquibase.util                           : -------------------------------
2024-08-27T17:40:42.613+03:00  INFO 6992 --- [           main] liquibase.util                           : Total change sets:            1
2024-08-27T17:40:42.614+03:00  INFO 6992 --- [           main] liquibase.util                           : Update summary generated
2024-08-27T17:40:42.644+03:00  INFO 6992 --- [           main] liquibase.lockservice                    : Successfully released change log lock
2024-08-27T17:40:42.647+03:00  INFO 6992 --- [           main] liquibase.command                        : Command execution complete
2024-08-27T17:40:42.770+03:00  INFO 6992 --- [           main] o.hibernate.jpa.internal.util.LogHelper  : HHH000204: Processing PersistenceUnitInfo [name: default]
2024-08-27T17:40:42.905+03:00  INFO 6992 --- [           main] org.hibernate.Version                    : HHH000412: Hibernate ORM core version 6.5.2.Final
2024-08-27T17:40:42.958+03:00  INFO 6992 --- [           main] o.h.c.internal.RegionFactoryInitiator    : HHH000026: Second-level cache disabled
2024-08-27T17:40:43.462+03:00  INFO 6992 --- [           main] o.s.o.j.p.SpringPersistenceUnitInfo      : No LoadTimeWeaver setup: ignoring JPA class transformer
2024-08-27T17:40:44.840+03:00  INFO 6992 --- [           main] o.h.e.t.j.p.i.JtaPlatformInitiator       : HHH000489: No JTA platform available (set 'hibernate.transaction.jta.platform' to enable JTA platform integration)
2024-08-27T17:40:44.893+03:00 ERROR 6992 --- [           main] j.LocalContainerEntityManagerFactoryBean : Failed to initialize JPA EntityManagerFactory: [PersistenceUnit: default] Unable to build Hibernate SessionFactory; nested exception is org.hibernate.tool.schema.spi.SchemaManagementException: Schema-validation: missing table [books]
2024-08-27T17:40:44.894+03:00  WARN 6992 --- [           main] ConfigServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'entityManagerFactory' defined in class path resource [org/springframework/boot/autoconfigure/orm/jpa/HibernateJpaConfiguration.class]: [PersistenceUnit: default] Unable to build Hibernate SessionFactory; nested exception is org.hibernate.tool.schema.spi.SchemaManagementException: Schema-validation: missing table [books]
2024-08-27T17:40:44.894+03:00  INFO 6992 --- [           main] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Shutdown initiated...
2024-08-27T17:40:44.907+03:00  INFO 6992 --- [           main] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Shutdown completed.
2024-08-27T17:40:44.911+03:00  INFO 6992 --- [           main] o.apache.catalina.core.StandardService   : Stopping service [Tomcat]
2024-08-27T17:40:44.929+03:00  INFO 6992 --- [           main] .s.b.a.l.ConditionEvaluationReportLogger : 

Error starting ApplicationContext. To display the condition evaluation report re-run your application with 'debug' enabled.
2024-08-27T17:40:44.954+03:00 ERROR 6992 --- [           main] o.s.boot.SpringApplication               : Application run failed

org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'entityManagerFactory' defined in class path resource [org/springframework/boot/autoconfigure/orm/jpa/HibernateJpaConfiguration.class]: [PersistenceUnit: default] Unable to build Hibernate SessionFactory; nested exception is org.hibernate.tool.schema.spi.SchemaManagementException: Schema-validation: missing table [books]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1806) ~[spring-beans-6.1.11.jar:6.1.11]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:600) ~[spring-beans-6.1.11.jar:6.1.11]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:522) ~[spring-beans-6.1.11.jar:6.1.11]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:337) ~[spring-beans-6.1.11.jar:6.1.11]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-6.1.11.jar:6.1.11]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:335) ~[spring-beans-6.1.11.jar:6.1.11]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:205) ~[spring-beans-6.1.11.jar:6.1.11]
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:954) ~[spring-context-6.1.11.jar:6.1.11]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:625) ~[spring-context-6.1.11.jar:6.1.11]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:146) ~[spring-boot-3.3.2.jar:3.3.2]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:754) ~[spring-boot-3.3.2.jar:3.3.2]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:456) ~[spring-boot-3.3.2.jar:3.3.2]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:335) ~[spring-boot-3.3.2.jar:3.3.2]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1363) ~[spring-boot-3.3.2.jar:3.3.2]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1352) ~[spring-boot-3.3.2.jar:3.3.2]
	at com.example.bookshop.SpringBootIntroApplication.main(SpringBootIntroApplication.java:10) ~[classes/:na]
Caused by: jakarta.persistence.PersistenceException: [PersistenceUnit: default] Unable to build Hibernate SessionFactory; nested exception is org.hibernate.tool.schema.spi.SchemaManagementException: Schema-validation: missing table [books]
	at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.buildNativeEntityManagerFactory(AbstractEntityManagerFactoryBean.java:421) ~[spring-orm-6.1.11.jar:6.1.11]
	at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.afterPropertiesSet(AbstractEntityManagerFactoryBean.java:396) ~[spring-orm-6.1.11.jar:6.1.11]
	at org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean.afterPropertiesSet(LocalContainerEntityManagerFactoryBean.java:366) ~[spring-orm-6.1.11.jar:6.1.11]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1853) ~[spring-beans-6.1.11.jar:6.1.11]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1802) ~[spring-beans-6.1.11.jar:6.1.11]
	... 15 common frames omitted
Caused by: org.hibernate.tool.schema.spi.SchemaManagementException: Schema-validation: missing table [books]
	at org.hibernate.tool.schema.internal.AbstractSchemaValidator.validateTable(AbstractSchemaValidator.java:135) ~[hibernate-core-6.5.2.Final.jar:6.5.2.Final]
	at org.hibernate.tool.schema.internal.GroupedSchemaValidatorImpl.validateTables(GroupedSchemaValidatorImpl.java:46) ~[hibernate-core-6.5.2.Final.jar:6.5.2.Final]
	at org.hibernate.tool.schema.internal.AbstractSchemaValidator.performValidation(AbstractSchemaValidator.java:98) ~[hibernate-core-6.5.2.Final.jar:6.5.2.Final]
	at org.hibernate.tool.schema.internal.AbstractSchemaValidator.doValidation(AbstractSchemaValidator.java:76) ~[hibernate-core-6.5.2.Final.jar:6.5.2.Final]
	at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.performDatabaseAction(SchemaManagementToolCoordinator.java:289) ~[hibernate-core-6.5.2.Final.jar:6.5.2.Final]
	at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.lambda$process$5(SchemaManagementToolCoordinator.java:144) ~[hibernate-core-6.5.2.Final.jar:6.5.2.Final]
	at java.base/java.util.HashMap.forEach(HashMap.java:1421) ~[na:na]
	at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.process(SchemaManagementToolCoordinator.java:141) ~[hibernate-core-6.5.2.Final.jar:6.5.2.Final]
	at org.hibernate.boot.internal.SessionFactoryObserverForSchemaExport.sessionFactoryCreated(SessionFactoryObserverForSchemaExport.java:37) ~[hibernate-core-6.5.2.Final.jar:6.5.2.Final]
	at org.hibernate.internal.SessionFactoryObserverChain.sessionFactoryCreated(SessionFactoryObserverChain.java:35) ~[hibernate-core-6.5.2.Final.jar:6.5.2.Final]
	at org.hibernate.internal.SessionFactoryImpl.<init>(SessionFactoryImpl.java:322) ~[hibernate-core-6.5.2.Final.jar:6.5.2.Final]
	at org.hibernate.boot.internal.SessionFactoryBuilderImpl.build(SessionFactoryBuilderImpl.java:457) ~[hibernate-core-6.5.2.Final.jar:6.5.2.Final]
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.build(EntityManagerFactoryBuilderImpl.java:1506) ~[hibernate-core-6.5.2.Final.jar:6.5.2.Final]
	at org.springframework.orm.jpa.vendor.SpringHibernateJpaPersistenceProvider.createContainerEntityManagerFactory(SpringHibernateJpaPersistenceProvider.java:75) ~[spring-orm-6.1.11.jar:6.1.11]
	at org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean.createNativeEntityManagerFactory(LocalContainerEntityManagerFactoryBean.java:390) ~[spring-orm-6.1.11.jar:6.1.11]
	at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.buildNativeEntityManagerFactory(AbstractEntityManagerFactoryBean.java:409) ~[spring-orm-6.1.11.jar:6.1.11]
	... 19 common frames omitted
```

And also I get something like this in the db for the books table:
image

The other tables do not have such problems: 
image


i did rollbacks, could that have made any difference? I didn't find anything suitable to solve the problem

**Comment by BelitskyiAlexandr** [comment]
I changed to "create-drop" because I can work with it at least

File: src/main/java/com/example/bookshop/mapper/BookMapper.java [hunk 3]
```diff
@@ -11,4 +11,30 @@ public interface BookMapper {
     BookDto toBookDto(Book book);
 
     Book toModel(CreateBookRequestDto bookDto);
+
+    default Book updateBookFromDto(CreateBookRequestDto createBookRequestDto, Book book) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/mapper/BookMapper.java:15]:
```suggestion
    void updateBookFromDto(CreateBookRequestDto createBookRequestDto, @MappingTarget Book entity);
```

File: src/main/java/com/example/bookshop/mapper/BookMapper.java [hunk 4]
```diff
@@ -11,4 +11,30 @@ public interface BookMapper {
...
+        }
+        if (createBookRequestDto.getDescription() != null) {
+            book.setDescription(createBookRequestDto.getDescription());
+        }
+        if (createBookRequestDto.getCoverImage() != null) {
+            book.setCoverImage(createBookRequestDto.getCoverImage());
+        }
+
+        return book;
+    }
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookshop/mapper/BookMapper.java:39]:
```suggestion
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Please, ask your question in chat to provide more details about your problem

File: src/main/java/com/example/bookshop/model/Book.java [hunk 5]
```diff
@@ -5,14 +5,20 @@
...
 import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
 
-@Entity
 @Getter
 @Setter
+@Entity
+@SQLDelete(sql = "UPDATE books SET is_deleted = true WHERE id=?")
+@SQLRestriction(value = "is_deleted=false")
```
Comment (okuzan) [inline src/main/java/com/example/bookshop/model/Book.java:20]:
```suggestion
@SQLDelete(sql = "UPDATE books SET is_deleted = TRUE WHERE id = ?")
@SQLRestriction(value = "is_deleted = FALSE")
```

File: src/main/resources/db/changelog/changes/01-create-book-table.yaml [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+                  name: description
+                  type: varchar(255)
+              - column:
+                  name: cover_image
+                  type: varchar(255)
+              - column:
+                  name: is_deleted
+                  type: bit(1)
+                  constraints:
+                    nullable: false
```
Comment (okuzan) [inline src/main/resources/db/changelog/changes/01-create-book-table.yaml:47]:
```suggestion
                    nullable: false
                    
```

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 7]
```diff
@@ -0,0 +1,4 @@
+databaseChangeLog:
+  - include:
+      file: /db/changelog/changes/01-create-book-table.yaml
+
```
Comment (okuzan) [inline src/main/resources/db/changelog/db.changelog-master.yaml:4]:
```suggestion
```

**Review by okuzan — APPROVED** [review]

**Pull request merged**